[SPARK-37823][INFRA][TESTS] Add `is-changed.py` dev script

dongjoon-hyun opened the pull request:
### What changes were proposed in this pull request?

This PR aims to add `is-changed.py` dev script to reuse in CIs for various purposes in the future.
To do that, file change detection codes are moved into a new file, `dev/utils.py`.

```
$ dev/is-changed.py -h
usage: is-changed [-h] [-f] [-m MODULES]

optional arguments:
  -h, --help            show this help message and exit
  -f, --fail            Exit with 1 if there is no relevant change.
  -m MODULES, --modules MODULES
                        A comma-separated list of modules to test (default: avro,build,catalyst,core,docker-integration-tests,docs,examples,graphx,hadoop-cloud,hive,hive-
                        thriftserver,kubernetes,kvstore,launcher,mesos,mllib,mllib-local,network-common,network-shuffle,pyspark-core,pyspark-ml,pyspark-mllib,pyspark-pandas,pyspark-pandas-slow,pyspark-
                        resource,pyspark-sql,pyspark-streaming,repl,root,sketch,spark-ganglia-lgpl,sparkr,sql,sql-kafka-0-10,streaming,streaming-kafka-0-10,streaming-kinesis-asl,tags,unsafe,yarn)
```

### Why are the changes needed?

This can be used in CI to save the community resources.

### Does this PR introduce _any_ user-facing change?

No, this is a dev-only test script change.

### How was this patch tested?

This should pass the existing CIs.

The following is a manual test.
```
$ APACHE_SPARK_REF=HEAD dev/is-changed.py         
false

$ APACHE_SPARK_REF=08dd010860 dev/is-changed.py        
true
```

This also supports `module` arguments to check the required modules exactly.
```
$ git diff HEAD~1       
diff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
index 252dd5bad3..e415ee05e2 100644
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
@@ -1,4 +1,5 @@
 /*
+
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
  * this work for additional information regarding copyright ownership.

$ APACHE_SPARK_REF=HEAD~1 dev/is-changed.py -m sql
true

$ APACHE_SPARK_REF=HEAD~1 dev/is-changed.py -m catalyst
true

$ APACHE_SPARK_REF=HEAD~1 dev/is-changed.py -m docs    
false
```

dongjoon-hyun committed:
[SPARK-37823][INFRA][TESTS] Add `is-changed.py` dev script

dongjoon-hyun commented on dev/run-tests.py:
```diff
@@ -23,145 +23,21 @@
 from sparktestsupport.toposort import toposort_flatten
 import sparktestsupport.modules as modules
 
 
```
This is moved into `dev/utils.py` to reuse it.

dongjoon-hyun commented:
Could you review this please, @HyukjinKwon ?

HyukjinKwon commented on dev/utils.py:
```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python3
```
What about moving this to under `dev/sparktestsupport`? That directory has some util-ish modules related to tests and CI.

dongjoon-hyun replied:
Sure!

HyukjinKwon commented on dev/is-changed.py:
```diff
@@ -0,0 +1,70 @@
+#
+
+from argparse import ArgumentParser
+from utils import *
```
Maybe avoiding whildcard import would be good. I think Python linter would complain anyway ..

HyukjinKwon commented on dev/run-tests.py:
```diff
@@ -23,145 +23,21 @@
 import sparktestsupport.modules as modules
 
 
-# -------------------------------------------------------------------------------------------------
```
One thing is that the doctests in `dev/utils.py ` wouldn't be executed (because the script executes the test before each invokation - https://github.com/apache/spark/blob/ea14cb92a146bcd6fa208e9395da1051c037e64a/dev/run-tests.py#L631)

HyukjinKwon reviewed (commented):
For separating the change detection logic, I think it's a good idea.
If this script is planned to be used in Apache Spark CI in the future, adding script sounds good too.

dongjoon-hyun committed:
Address comments

dongjoon-hyun commented:
Could you review this again, @HyukjinKwon ? I addressed all comments.

- Move to `sparktestsupport` folder.
- Fix all Python linter issues including removing `import *`.
- Check `utils.py` doctest inside `sparktestsupport`.

```bash
$ python3 -m doctest -v sparktestsupport/utils.py | tail -n11
3 items had no tests:
    utils
    utils._test
    utils.determine_tags_to_exclude
3 items passed all tests:
   3 tests in utils.determine_modules_for_files
   8 tests in utils.determine_modules_to_test
   2 tests in utils.identify_changed_files_from_git_commits
13 tests in 6 items.
13 passed and 0 failed.
Test passed.
```

dongjoon-hyun commented:
All tests passed except the irrelevant `ImageFileFormatSuite`'s three failures which are known flaky tests.
```
[info] *** 3 TESTS FAILED ***
[error] Failed: Total 1760, Failed 3, Errors 0, Passed 1757, Ignored 7
[error] Failed tests:
[error] 	org.apache.spark.ml.source.image.ImageFileFormatSuite
[error] (mllib / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 1455 s (24:15), completed Jan 6, 2022 6:21:44 AM
```

dongjoon-hyun committed:
empty commit

dongjoon-hyun committed:
Run doctest

HyukjinKwon commented on dev/sparktestsupport/utils.py:
```diff
@@ -0,0 +1,159 @@
+        if m not in changed_modules:
+            tags += m.test_tags
+    return tags
+
+
+def _test():
+    import doctest
+    failure_count = doctest.testmod()[0]
+    if failure_count:
+        sys.exit(-1)
+
+if __name__ == "__main__":
+    _test()
```
```suggestion
```

Maybe we can remove these because these are not supposed to be executed as a script.

HyukjinKwon commented on dev/is-changed.py:
```diff
@@ -0,0 +1,73 @@
+import os
+import sys
+from argparse import ArgumentParser
+from sparktestsupport.utils import determine_modules_for_files, determine_modules_to_test, identify_changed_files_from_git_commits
```
Hm, it actually should complain about the line length .. I actually fixed the linter script to run on dev scripts too before but seems like there's a regression about this. I will take a separate look at this one - there should be more places to fix per the linter not only this script.

dongjoon-hyun replied:
Got it~ Thanks.

HyukjinKwon replied:
Actually, the linter for line-length is disabled for the whole project after https://github.com/apache/spark/pull/34655 .. 😢 . let me make a fix soon after merging this one.

HyukjinKwon reviewed (approved):
LGTM

HyukjinKwon commented:
Merged to master.

dongjoon-hyun commented:
Thank you so much!
